bftengine/src/preprocessor/RequestProcessingInfo.cpp

```diff
@@ -10,30 +10,72 @@
 // file.
 
 #include "RequestProcessingInfo.hpp"
-#include "ReplicaConfig.hpp"
 #include "Logger.hpp"
 
 namespace preprocessor {
 
 using namespace std;
 using namespace concord::util;
-using namespace concordUtils;
 
-RequestProcessingInfo::RequestProcessingInfo(uint16_t numOfReplicas, ReqId reqSeqNum)
+uint16_t RequestProcessingInfo::numOfRequiredEqualReplies_ = 0;
+
+RequestProcessingInfo::RequestProcessingInfo(uint16_t numOfReplicas, uint16_t numOfRequiredReplies, ReqId reqSeqNum)
     : numOfReplicas_(numOfReplicas), reqSeqNum_(reqSeqNum) {
-  for (auto i = 0; i < numOfReplicas; i++)
-    // Placeholders for all replicas
-    replicasDataForRequest_.push_back(nullptr);
+  numOfRequiredEqualReplies_ = numOfRequiredReplies;
   LOG_DEBUG(GL, "Created RequestProcessingInfo with reqSeqNum=" << reqSeqNum_ << ", numOfReplicas= " << numOfReplicas_);
 }
 
-void RequestProcessingInfo::saveClientPreProcessRequestMsg(const ClientPreProcessReqMsgSharedPtr& clientPreProcessReq) {
-  clientPreProcessRequestMsg_ = clientPreProcessReq;
+void RequestProcessingInfo::handlePrimaryPreProcessed(PreProcessRequestMsgSharedPtr msg,
+                                                      const char *preProcessResult,
+                                                      uint32_t preProcessResultLen) {
+  numOfReceivedReplies_++;
+  preProcessRequestMsg_ = msg;
+  myPreProcessResult_ = preProcessResult;
+  myPreProcessResultLen_ = preProcessResultLen;
+  myPreProcessResultHash_ = convertToArray(SHA3_256().digest(myPreProcessResult_, myPreProcessResultLen_).data());
+  preProcessingResultHashes_[myPreProcessResultHash_]++;
+}
+
+void RequestProcessingInfo::handlePreProcessReplyMsg(PreProcessReplyMsgSharedPtr preProcessReplyMsg) {
+  numOfReceivedReplies_++;
+  preProcessingResultHashes_[convertToArray(preProcessReplyMsg->resultsHash())]++;  // Count equal hashes
+}
+
+HashArray RequestProcessingInfo::convertToArray(const uint8_t resultsHash[SHA3_256::SIZE_IN_BYTES]) {
+  HashArray hashArray;
+  for (uint64_t i = 0; i < SHA3_256::SIZE_IN_BYTES; i++) hashArray[i] = resultsHash[i];
+  return hashArray;
```

Comment:
why do we need this conversion?

Reply:
The simplest way to find the biggest number of same hashes is by counting them while inserting to the map/hash table. Technically you cannot define C++ map key as a static array, so I convert values to std::array.

Reply:
Instead of the copy, you may be able to store the std::array directly in the message.